db/event_manager_test.go

```diff
@@ -831,3 +831,69 @@ func assertChannelLengthWithTimeout(t *testing.T, c chan interface{}, expectedLe
 		}
 	}
 }
+
+func mockDBStateChangeEvent(dbName string, state string, reason string, adminInterface string) *DBStateChangeEvent {
+	body := make(Body, 5)
+	body["dbname"] = dbName
+	body["admininterface"] = adminInterface
+	body["state"] = state
+	body["reason"] = reason
+	body["localtime"] = time.Now().Format(base.ISO8601Format)
+	event := &DBStateChangeEvent{Doc: body}
+	return event
+}
+
+type UnsupportedEvent struct {
+	AsyncEvent
+}
+
+func (event *UnsupportedEvent) String() string {
+	return fmt.Sprintf("Couchbase Sync Gateway doesn't support this kind of events!")
+}
+
+func (event *UnsupportedEvent) EventType() EventType {
+	return EventType(255)
+}
+
+// Simulate the scenario for handling unsupported events.
+func TestWebhookHandleUnsupportedEvenType(t *testing.T) {
+	prev, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+	wh := &Webhook{url: "https://127.0.0.1:8086/webhook?"}
+	event := &UnsupportedEvent{}
+	wh.HandleEvent(event)
+	curr, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+	warns := curr - prev
+	assert.Equal(t, 1, warns)
+}
+
+// Simulate the filter function processing abort scenario.
+func TestWebhookHandleEventDBStateChangeFilterFuncError(t *testing.T) {
+	prev, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+	wh := &Webhook{url: "https://127.0.0.1:8086/couchbaseWebhook?"}
+	event := mockDBStateChangeEvent("db", "online", "Index service is listening", "127.0.0.1:4985")
+	source := `function (doc) { invalidKeyword if (doc.state == "online") { return true; } else { return false; } }`
+	wh.filter = NewJSEventFunction(source)
+	wh.HandleEvent(event)
+	curr, err := strconv.Atoi(base.StatsResourceUtilization().Get(base.StatKeyWarnCount).String())
+	assert.NoError(t, err, "No error while getting the count of warnings")
+	warns := curr - prev
+	assert.Equal(t, 2, warns, "Filter function processing should be aborted and warnings should be logged")
+}
+
+// Simulate marshalling doc error for webhook post against DBStateChangeEvent
+func TestWebhookHandleEventDBStateChangeMarshallDocError(t *testing.T) {
```

Comment:
Typo in Marshal

Reply:
This has been fixed and changes are available in the latest revision.